### PR TITLE
Allow string keys

### PR DESCRIPTION
**Pull request by chawes13**
## Items Addressed
- Addresses #99 

## Author Checklist
- [x] Add unit test(s)
- [x] ~Update documentation (if necessary)~
- [x] Update version in `version.rb` following [versioning guidelines](https://github.com/LaunchPadLab/opex/blob/master/gists/gem-guidelines.md#pull-requests-and-deployments)

### CODE REVIEW

File: lib/decanter/core.rb [hunk 1]
```diff
@@ -210,7 +210,11 @@ def strict_mode
       def value_missing?(value)
         value.nil? || value == ""
       end
-
+      
+      def transform_args(args)
```
Comment (dpikt) [inline lib/decanter/core.rb]: Add a comment about what this method does?

File: lib/decanter/core.rb [hunk 2]
```diff
@@ -9,7 +9,7 @@ module ClassMethods
       
       def input(name, parsers=nil, **options)
 
-        _name = [name].flatten
+        _name = [name].flatten.map(&:to_sym)
```
Comment (dpikt) [inline lib/decanter/core.rb]: Add a comment about what this line does?

**Comment by chawes13** [comment]
@dpikt What are your thoughts on requiring input names to be symbols? E.g.,
```rb
class TripDecanter < Decanter::Base
  input :first_name, :string # works
  input 'last_name', :string # raises error
end
```
Would that follow convention compared to other gems or should we retain the flexibility of allowing either (using `with_indifferent_access`)?

**Comment by dpikt** [comment]
I think we should stick with whatever the current behavior is on that, since it's probably not worth the breaking change. If I were writing the library from scratch, I would probably require them to be symbols.

**Comment by chawes13** [comment]
> I think we should stick with whatever the current behavior is on that, since it's probably not worth the breaking change. If I were writing the library from scratch, I would probably require them to be symbols.

@dpikt I think the current behavior has a defect (see below). Maybe the answer is to close this PR and fix the root issue before introducing `with_indifferent_access` for all args in an upcoming major release?

```rb
class TripDecanter < Decanter::Base
  input 'description', :string
end
```
## Current Behavior
- String input gets flagged as unhandled in strict mode for `ActionControl::Parameters`
- String input gets flagged as unhandled in strict mode for a ruby hash with symbol key
- String input gets flagged as unhandled in strict mode for a ruby hash with string key

## Expected Behavior
- String input gets flagged as unhandled in strict mode for `ActionControl::Parameters`
- String input gets flagged as unhandled in strict mode for a ruby hash with symbol key
- String input is handled for a ruby hash with string key

**Comment by dpikt** [comment]
@chawes13 I'm getting mixed up on the distinction between the input declarations:
```rb
input 'description', :string
# vs
input :description, :string
```
and keys:
```rb
{
   description: 'foo'
}
# vs
{
   'description' => 'foo'
}
```

I think expected behavior for me would be if the input declaration type *didn't* affect the behavior of the decanter. That is, all of the above options would be valid and interoperable. Is that what this PR does?

**Comment by chawes13** [comment]
@dpikt Yes, it makes it so `input 'description'` and `input :description` are interoperable. I suppose since the current behavior considers all string input names as unhandled, it wouldn't be a breaking change.

**Comment by dpikt** [comment]
So then:
* String input is handled in strict mode for ActionController::Parameters
* String input is handled in strict mode for a ruby hash with symbol key
* String input is handled for a ruby hash with string key

Unless "strict" were redefined to do this type checking as well (I think right now it just dictates whether an exception is thrown)

**Comment by chawes13** [comment]
@dpikt bump 🙂

File: spec/decanter/decanter_core_spec.rb [hunk 3]
```diff
@@ -564,6 +564,45 @@ def self.name
           expect{ decanter.decant({ name: [] }) }
             .not_to raise_error
         end
+        it 'should treaty empty strings as missing' do
```
Comment (dpikt) [inline spec/decanter/decanter_core_spec.rb:567]: Typo

File: spec/decanter/decanter_core_spec.rb [hunk 4]
```diff
@@ -564,6 +564,45 @@ def self.name
...
+
+      context 'when params keys are strings' do
+        let(:decanter) {
+          Class.new(Decanter::Base) do
+            input :name, :string
+            input :description, :string
+          end
+        }
+        let(:args) { { 'name' => 'My Trip', 'description' => 'My Trip Description' } }
+        it 'returns the merged result' do
```
Comment (dpikt) [inline spec/decanter/decanter_core_spec.rb:585]: This assertion is a little vague - can you clarify what this is checking?

**Review by dpikt — APPROVED** [review]
Code looks good! Added a non-blocking suggestion on the spec copy.

**Pull request merged**